### PR TITLE
Create a new directory when a new file ends with `/`

### PR DESCRIPTION
**Pull request by 80avin**
This feature is present in almost all editors but in zed, doing so crashes the editor.

Release Notes:

- When creating new file via project panel, a filename ending with ``/`` creates a directory.

### CODE REVIEW

**Comment by 80avin** [comment]
@cla-bot check

File: crates/project_panel/src/project_panel.rs [hunk 1]
```diff
@@ -694,10 +694,12 @@ impl ProjectPanel {
 
         let worktree_id = edit_state.worktree_id;
         let is_new_entry = edit_state.is_new_entry;
+        let filename = self.filename_editor.read(cx).text(cx);
+        edit_state.is_dir =
+            edit_state.is_dir || (edit_state.is_new_entry && filename.ends_with('/'));
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:699]:
```suggestion
            edit_state.is_dir || (edit_state.is_new_entry && filename.ends_with(std::path::MAIN_SEPARATOR));
```

Let's not forget about Windows.

File: crates/project_panel/src/project_panel.rs [hunk 2]
```diff
@@ -2845,6 +2847,84 @@ mod tests {
         );
     }
 
+    #[gpui::test(iterations = 10)]
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs:2850]:
```suggestion
    #[gpui::test]
```

`iterations` are used to run the same test multiple times with different deterministic executor's seed, fuzzing the bugs related to e.g. races around future/task spawns.

We can run this test just once as it seems from its contents — there's no concurrency or hard logic in play?

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, looks like a nice improvement overall.

What would happen if we try to create a directory that already exists? Maybe worth testing that too?

**Comment by bbb651** [comment]
Worth noting vscode let's you create nested folders and optionally a file at the end, things like:
- `foo/bar`
- `foo/bar/`
- `foo/bar/baz`
and so on, which is a lot more complicated because it's ok for some prefix of the path to already exist.
I think I actually found a vscode bug around this behavior, with in the folder `foo`, it won't let you create the file `foo`:
image
But with the folder `foo` and the file `bar` inside it, it won't show anything wrong creating `foo/bar`, but will fail to do so:
image
Could always be added in future prs of course.

File: crates/project_panel/src/project_panel.rs (hunk 2 above)
Comment (80avin) [reply crates/project_panel/src/project_panel.rs:2850]: I didn't understood it at first. So just copied this test which looked similar

https://github.com/80avin/zed/blob/d0a971986925157f284507190ed20a1d73ffaf7e/crates/project_panel/src/project_panel.rs#L2717

File: crates/project_panel/src/project_panel.rs (hunk 1 above)
Comment (80avin) [reply crates/project_panel/src/project_panel.rs:699]: I cannot test it on windows but shouldn't it be `filename.ends_with(std::path::MAIN_SEPARATOR) || filename.ends_with('/')` then ?

The question is that if we allow both `/` and `\` as path separators on windows or only `\` ?

**Comment by 80avin** [comment]
>   I think I actually found a vscode bug around this behavior, with in the folder `foo`, it won't let you create the file `foo`:
>   image

I think you are creating the file at same level as `foo` directory that's why this issue is coming.
To create a file inside a directory, you have to select the directory first. It'll look like this
image


>   But with the folder `foo` and the file `bar` inside it, it won't show anything wrong creating `foo/bar`, but will fail to do so:

Noticed that in zed, even without this PR, it fails silently. We may need error for that but that can be separate PR.

File: crates/project_panel/src/project_panel.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/project_panel/src/project_panel.rs:699]: It's just `std::path::MAIN_SEPARATOR` which will be different depending on the platform it's compiled to.

**Review by SomeoneToIgnore — APPROVED** [review]
Thank you, let's reformat and merge this.

https://github.com/zed-industries/zed/pull/12196 should handle other creation bugs mentioned, so I think we'll soon be good with the directory creation in project panel.

File: crates/project_panel/src/project_panel.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/project_panel/src/project_panel.rs]: No tests are run on Windows yet, and I expect there are bugs.

> `filename.ends_with(std::path::MAIN_SEPARATOR) || filename.ends_with('/')` 

is never right, as `std::path::MAIN_SEPARATOR` is already `\` when compiled to Unix, and `/` to Windows targets.
Eventually, we'll have to fix all path-related issues and support both, but without direct manipulation of `/` or `\` symbols.

**Comment by bbb651** [comment]
> I think you are creating the file at same level as `foo` directory that's why this issue is coming. To create a file inside a directory, you have to select the directory first.

It was intentional, I gave it as an example of what trying to create a duplicate file should do, rather than letting you create the file and get the filesystem error

**Pull request merged**